src/VecSim/utils/serializer.cpp

```diff
@@ -0,0 +1,32 @@
+#include <fstream>
+#include <string>
+
+#include "VecSim/utils/serializer.h"
+
+// Persist index into a file in the specified location.
+void Serializer::saveIndex(const std::string &location) {
+
+    // Serializing with V2.
+    EncodingVersion version = EncodingVersion_V2;
+
+    std::ofstream output(location, std::ios::binary);
+    writeBinaryPOD(output, version);
+    saveIndexIMP(output);
+    output.close();
+}
+
+Serializer::EncodingVersion Serializer::ReadVersion(std::ifstream &input) {
+
+    input.seekg(0, std::ifstream::beg);
+
+    // The version number is the first field that is serialized.
+    EncodingVersion version = EncodingVersion_INVALID;
+    readBinaryPOD(input, version);
+    // Only V1 and V2 are supported
+    if (version >= EncodingVersion_INVALID) {
```

Comment:
you're right @meiravgri. But we don't return an error if `version==0`. You can just add this validation to this `if`

Reply:
EncodingVersion==0 won't throw because it is smaller than EncodingVersion_INVALID
`    typedef enum EncodingVersion {
        EncodingVersion_V1 = 1,
        EncodingVersion_V2 = 2,
        EncodingVersion_INVALID, // This should always be last.
    } EncodingVersion;`
I can also add EncodingVersion_V0,
but it will return when we try to initialize params. So I don't quite understand how to handle v0 without returning an error

Reply:
I think you SHOULD return an error for `version==0`

Reply:
Agrees. 
```suggestion
    // Only V1 and V2 are supported
    if (version <= 0 || version >= EncodingVersion_INVALID) {
```

Reply:
oh got it :sweat_smile: